batchflow/pipeline_executor.py

```diff
@@ -0,0 +1,321 @@
+        self.pipeline = pipeline
+        self.notifier = None
+
+        self._stop_flag = False
+        self._executor = None
+        self._service_executor = None
+        self._prefetch_count = None
+        self._prefetch_queue = None
+        self._batch_queue = None
+
+    def _clear_queue(self, queue):
+        if queue is not None:
+            while not queue.empty():
+                queue.get(block=True)
+                queue.task_done()
+
+    def _stop_executor(self, executor):
+        if executor is not None:
+            executor.shutdown(wait=False, cancel_futures=True)
+
+    def reset(self):
+        """ Clear all iteration metadata in order to start iterating from scratch """
+        self._stop_flag = True
+
+        self._clear_queue(self._prefetch_count)
+        self._clear_queue(self._prefetch_queue)
+        self._clear_queue(self._batch_queue)
+
+        self._stop_executor(self._executor)
+        self._stop_executor(self._service_executor)
+
+        self._executor = None
+        self._service_executor = None
+        self._prefetch_count = None
+        self._prefetch_queue = None
+        self._batch_queue = None
+
+
+    def _put_batches_into_queue(self, gen_batch, seed):
+        iteration = 1
+        while not self._stop_flag:
+            # this will block if there are too many batches prefetched
+            self._prefetch_count.put(1, block=True)
+            try:
+                batch = next(gen_batch)
+            except StopIteration:
+                break
+            else:
+                future = self._executor.submit(self.pipeline.execute_for, batch,
+                                               iteration=iteration, seed=seed.spawn(1)[0],
+                                               new_loop=True)
+                self._prefetch_queue.put(future, block=True)
+                iteration = iteration + 1
+        self._prefetch_queue.put(END_PIPELINE, block=True)
+
+    def _run_batches_from_queue(self, ignore_exceptions):
+        while not self._stop_flag:
+            future = self._prefetch_queue.get(block=True)
+
+            try:
+                if future == END_PIPELINE:
+                    batch = END_PIPELINE
+                    break
+                batch = future.result()
+
+            except SkipBatchException:
+                pass
+            except StopPipeline:
+                batch = END_PIPELINE
+                break
+            except Exception as exc:   # pylint: disable=broad-except
+                print("Exception:", exc)
+                traceback.print_tb(exc.__traceback__)
+                if ignore_exceptions:
+                    batch = None
+                else:
+                    batch = END_PIPELINE
+            finally:
+                self._prefetch_queue.task_done()
+                self._batch_queue.put(batch, block=True)
+
+
+    def gen_batch(self, *args, dataset=None, rebatch=False, reset='iter', profile=False,
```

Comment:
As this method (as well as the `PipelineExecutor` itself are purely internal constructions, do we need such elaborated docstring (which is, in any case, just a copy-and-paste from `Pipeline.get_batch`) here?